common/sampling.cpp

```diff
@@ -0,0 +1,166 @@
+#include "sampling.h"
+
+llama_sampling_state::~llama_sampling_state() {
+    for (auto & it : sequence_states) {
+        if (it.second.grammar != NULL) {
+            llama_grammar_free(it.second.grammar);
+            it.second.grammar = NULL;
+        }
+    }
+}
+
+llama_sampling_state llama_sampling_state_init(
+        const struct gpt_params & params,
+                  llama_grammar * grammar) {
+  llama_sampling_state result;
+
+  result.params = params.sampling_params;
+  result.grammar = grammar;
+  return result;
+}
+
+// Note: Creates the state if it doesn't exist, so this always return something.
+llama_sampler_sequence_state & llama_sampling_get_sequence_state(
+              llama_sampling_state & state,
+        const llama_seq_id           seq) {
+    const auto it = state.sequence_states.find(seq);
+    if (it != state.sequence_states.end()) {
+        return it->second;
+    }
+    llama_sampler_sequence_state new_state = {
+        2.0f * state.params.mirostat_tau,
+        state.grammar != NULL ? llama_grammar_copy(state.grammar) : NULL,
+    };
+    return state.sequence_states.insert({seq, new_state}).first->second;
+}
+
+bool llama_sampling_state_reset(
+              llama_sampling_state & state,
+        const llama_seq_id           seq) {
+    const auto it = state.sequence_states.find(seq);
+    if (it == state.sequence_states.end()) return false;
+    if (it->second.grammar != NULL) {
+        llama_grammar_free(it->second.grammar);
+        it->second.grammar = NULL;
+    }
+    state.sequence_states.erase(it);
+    return true;
+}
+
+llama_token llama_sample_token(
+                  struct llama_context * ctx,
+                  struct llama_context * ctx_guidance,
+                  struct llama_sampling_state & state,
+        const std::vector<llama_token> & last_tokens,
+         std::vector<llama_token_data> & candidates,
```

Comment:
Should we absorb `last_tokens` and `candidates` into `llama_sampling_state`?

Reply:
Last tokens can be specific to the sequence, right? So this would kind of mean stuff using last tokens would have to be aware of sequences. I kind of feel like this also might limit how people can manipulate last tokens and if there are currently examples that do that kind of thing it might be difficult to adapt them (for me anyway, since I'm not really deeply familiar with most of them). 

`candidates` I'm less sure about, it's basically just a scratch area for the logits in a form samplers can work with (right?) so I think moving it in there is less of a big deal. It's a pretty large structure though, I don't know if that's a consideration. Right now the current stuff in those structs is pretty lightweight.

I don't have very strong feelings about this. I'd like to say "These changes are already complicated enough, let's come back to that" but... I probably never would. :)